### PR TITLE
Parameter Filtration in LM Optimizer

### PR DESCRIPTION
**Pull request by LeonOtis**
## Proposed changes

This is the remainder of the changes originally proposed in #4061 . This PR allows the wave function parameters optimized by the linear method to be filtered based on the amount of noise in the parameter gradients. The LM will then optimize only a subset of the parameters while the others can be handled by accelerated descent in a hybrid optimization. 

The main changes are in the LM engine and the QMCFixedSampleLinearOptimizeBatched driver. There are a few additional user inputs for turning on parameter filtration, which are currently being handled similarly to the many other driver inputs. The parameter filtration should now be compatible with both MPI and OMP parallelism. Thanks again for all the feedback so far and I am always happy to make further improvements.

## What type(s) of changes does this code introduce?

- New feature

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?

Lawrencium computing cluster

## Checklist

- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- No. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- Yes. Documentation has been added (if appropriate)

### CODE REVIEW

**Comment by markdewing** [comment]
Test this please

File: src/QMCDrivers/WFOpt/EngineHandle.h
Comment (PDoakORNL) [inline src/QMCDrivers/WFOpt/EngineHandle.h:133]: I'd encourage you to read the section on [thread safety of stl containers](https://en.cppreference.com/w/cpp/container)

File: src/QMCDrivers/tests/test_EngineHandle.cpp
Comment (PDoakORNL) [inline src/QMCDrivers/tests/test_EngineHandle.cpp:50]: this needs to be tested in a threaded context.

Comment (PDoakORNL) [reply src/QMCDrivers/tests/test_EngineHandle.cpp:50]: in fact all of the engine methods need to be or it needs to be.  Or they need to be called only from one thread.

File: src/QMCDrivers/WFOpt/EngineHandle.h
Comment (PDoakORNL) [inline src/QMCDrivers/WFOpt/EngineHandle.h:154]: There looks to be a potential disaster data race above this on der_rat_samp and le_der_samp.

The only id you should care about in a threaded section is the crowd id, it may or may not have any relation to the omp thead id. Crowds may be called at the same time on any collection of thread_id's the concurrency context you're in is identified by the crowd id and you should avoid needing to know it.

**Review by PDoakORNL — CHANGES_REQUESTED** [review]
Unless I'm reading this very wrong there are serious fundamental problems with the concurrency logic of the batched linear optimization driver and this PR.  Please mark this as [WIP] until the concurrency issues in it are fixed.

One of the maintainers should help you work through this in detail.

Have you run this with opt_num_crowds > 1?  I would expect something interesting to happen.

File: src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp
Comment (markdewing) [inline src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp:295]: It would be better to keep the arrays in LMYEngine sized as the full number of samples (per rank, in rank_local_num_samples_) and use samples_per_crowd_offsets to index into those arrays.   That should get rid of all (or most) of the thread calls in takeSample.

Comment (markdewing) [reply src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp:295]: takeSample has `ib` passed an argument, which is the crowd-local index.  It needs to be updated to also pass `is` ( supposed to be short for 'Index into Samples', not the word `is`), which is the index into samples on each rank.

Comment (PDoakORNL) [reply src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp:295]: Both of those argument names could be long enough to carry information.  Two letters excepting those that are common words 'is', 'as', etc are fine for an iterated index in a small loop but not great as argument names.  `local_index` and `index_offset` come to mind.  Then only index offset would need non trivial docs.

Comment (PDoakORNL) [reply src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp:295]: And I think sizing the arrays to the full number of samples outside of the concurrent lambda would do a good deal here to get a hacked first version, but every method called in the concurrent context and its calls etc needs to be evaluated for thread safety.  It's quite likely they will work fine in a single threaded context.

**Comment by LeonOtis** [comment]
Thanks for the feedback. The new set of changes have the history arrays in the LM engine sized according to the number of samples on an MPI rank and different threads should store data in them using the index offsets. Tests I've done with multiple threads have worked, but there could be cases I've overlooked. I'm happy to do further rounds of fixes as needed.

File: src/QMCDrivers/WFOpt/EngineHandle.h
Comment (markdewing) [inline src/QMCDrivers/WFOpt/EngineHandle.h:49]: I'm not sure this is best described as an offset. Maybe "sample_index"?  This is the index of the sample on this MPI rank.   The extent of the index is [0, rank_local_num_samples_), so maybe rank_local_sample_index?  The rank_local part is a bit wordy, so sample_index should be sufficient.

Comment (LeonOtis) [reply src/QMCDrivers/WFOpt/EngineHandle.h]: Ok, I have changed the naming to sample_index to make things clearer.

**Comment by markdewing** [comment]
Do you have an example input file for this feature?  Ideally there should be a test added that also serves as an example.  However, parameter filtering gets used when scaling up the number of parameters than can be handled and a test that is small enough to be a good test might not be a good example.  It also sounds like parameter filtering gets used in combination with hybrid optimization.  it would be good to have a more complete example to see how the feature works in use.  Attaching files to the PR is good enough for now.

**Comment by LeonOtis** [comment]
[example_input.zip](https://github.com/QMCPACK/qmcpack/files/9244172/example_input.zip)

Sure, here are input files for a quick example with the carbon trimer. The actual optimization results won't be very meaningful with low sample numbers and filtering an already small number of parameters, but the example_filter_hybrid.xml file shows the new inputs in the context of a hybrid optimization. Parameter filtration can be used for the LM on its own just by using the LM section of the hybrid input as a standalone, but I would generally expect the results to be inferior to allowing descent to optimize any parameters the LM leaves alone. I think this PR should already have some changes to the documentation to explain the new inputs, but I can add more if needed.

**Comment by LeonOtis** [comment]
Added a small fix to the legacy driver to correct a problem @ye-luo noticed with PR #4075. At this point, what else is needed before this PR can be merged? I'm happy to revise the handling of threading more if desired. I'm also willing to come up with tests if needed though I wanted to be sure the threading issues were settled before trying to design any.

**Comment by ye-luo** [comment]
@LeonOtis could you make a new branch from develop and cherry pick the last fix commit you pushed and then make a separate PR.

**Comment by LeonOtis** [comment]
@ye-luo Ok, I've opened PR #4161 for that small change.

**Comment by LeonOtis** [comment]
I've now added a unit test for parameter filtration. Are there any other changes I should make before this PR can be merged? @ye-luo @markdewing

**Comment by ye-luo** [comment]
@LeonOtis I will make a review of this PR within this week. Q: with the current develop, does the "hybrid" method working with batched driver? When you made the fix to the legacy driver, I noticed there were no corresponding tests for the batched driver. If the current develop already works with "hybrid" and batched driver, could you introduce some tests.

**Comment by LeonOtis** [comment]
@ye-luo The hybrid method should work with the batched driver. I've now added a test for it on H4, similar to what I previously added for the legacy driver.

**Comment by ye-luo** [comment]
> @ye-luo The hybrid method should work with the batched driver. I've now added a test for it on H4, similar to what I previously added for the legacy driver.

Thank you. Could you make a separate PR with the new commits to develop? Since it works with the develop already.

**Comment by LeonOtis** [comment]
> > @ye-luo The hybrid method should work with the batched driver. I've now added a test for it on H4, similar to what I previously added for the legacy driver.
> 
> Thank you. Could you make a separate PR with the new commits to develop? Since it works with the develop already.

Ok, I've opened a new PR #4172 that just adds the test.

File: src/QMCDrivers/tests/test_EngineHandle.cpp
Comment (ye-luo) [reply src/QMCDrivers/tests/test_EngineHandle.cpp:50]: It prepares storage for each sample.
This is called outside the thread context and thread agnostic.

File: src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.cpp
Comment (ye-luo) [inline src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.cpp:694]: Why not m_param.add directly filter_param_, filter_info_, store_samples_?

File: src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.cpp
Comment (ye-luo) [inline src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.cpp:1194]: Remove?

File: src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.h
Comment (ye-luo) [inline src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.h:255]: Could you put all the LMY specific option variables in a structure?

File: src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp
Comment (ye-luo) [inline src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp:372]: Could you adjust the takeSample API and move the call out of the ib loop?
takeSample(energy_list, dlogpsi_array, dhpsioverpsi_array, base_sample_index);
In the takeSample implementation, handle the ib loop.

**Review by ye-luo — COMMENTED** [review]
@LeonOtis I did some cleaning, please pull before continuing editing. Once my requests are addressed, I think this PR can get in. I close my eyes on run_three_shift and everything inside formic.

File: src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp
Comment (LeonOtis) [reply src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp]: Ok, I've put the ib loop inside takeSample.

File: src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.cpp
Comment (LeonOtis) [reply src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.cpp]: Removed.

File: src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.cpp
Comment (LeonOtis) [reply src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.cpp:694]: I find when I try m_param.add directly for those booleans, the user input must be "1" to for them to be set to true. Using the strings seemed to allow for more intuitive user input and follows what's done for the blocked LM and excited state settings.

File: src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.h
Comment (LeonOtis) [reply src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.h]: I've grouped the optimizer input parameters together in an unnamed struct. I had to leave the string parameters outside the struct to compile successfully, but they are listed just above the struct if this change is just for organizing the file.

File: src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.cpp
Comment (ye-luo) [reply src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.cpp:694]: Thanks. I will keep a note to update bool handling.

File: src/QMCDrivers/WFOpt/EngineHandle.h
Comment (ye-luo) [inline src/QMCDrivers/WFOpt/EngineHandle.h:87]: I think we should avoid using number of threads and thread id.

File: src/QMCDrivers/WFOpt/EngineHandle.h
Comment (ye-luo) [inline src/QMCDrivers/WFOpt/EngineHandle.h:111]: would like to move away from thread_num

**Review by ye-luo — DISMISSED** [review]
There is still an issue of explicitly using threads. That is not a good pattern. However, I'd like that being addressed in a later PR.

**Comment by ye-luo** [comment]
Ping @PDoakORNL @markdewing since you gave review comments as well.

File: src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.cpp
Comment (ye-luo) [reply src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.cpp:694]: I actually fixed the ParameterSet handling boolean in the develop.
@PDoakORNL seems out of office. As we are waiting for his review, could you try removing the input strings?

Comment (LeonOtis) [reply src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.cpp]: Ok, thanks. Removing the strings works now and I've pushed a commit.

File: src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.h
Comment (PDoakORNL) [inline src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.h:203]: the naming of all these private variables does not follow the coding standards of QMCPACK.

File: src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.h
Comment (PDoakORNL) [inline src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.h:206]: What purpose does this anonymous struct serve?  
To my knowledge anonymous structs are not part of the c++ standard although I guess they compile for most compilers.
Does this make the code easier to read or better in anyway?  If you want a nested data structure name it, declare a class member of that nested type with a name following the naming convention i.e. `lmy_in_` and access it properly for example `lmy_in_.num_shifts`

File: src/QMCDrivers/WFOpt/EngineHandle.h
Comment (PDoakORNL) [inline src/QMCDrivers/WFOpt/EngineHandle.h:112]: How this relates to all the replica based accesses in the DescentEngine itself is really cryptic. Since this oenpmp thread id which isn't guaranteed to be any particular number especially in the batched context yet it is used as the replica id in the engine, which seems like it is very important.

File: src/QMCDrivers/WFOpt/EngineHandle.h
Comment (PDoakORNL) [inline src/QMCDrivers/WFOpt/EngineHandle.h:85]: num samples is unused.

**Review by PDoakORNL — CHANGES_REQUESTED** [review]
I think some of the thread safety issues I previously was worried have been cleaned up but the subtypes of EngineHandle seem to be trouble and perhaps only one is really relevant to this PR.

DescentEngineHandle/Descent engine seems incoherent, if this is just here to get things compiling keep the unfinished code in your private branch and remove it from the PR throw a runtime exception if anyone tries to use it.

That would get us to smaller issues like the anonymous struct.

File: src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp
Comment (ye-luo) [reply src/QMCDrivers/WFOpt/QMCCostFunctionBatched.cpp:281]: I'm not getting this part "don't infact have the same sematics for the prepareSampling and takeSample Methods."

File: src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.h
Comment (ye-luo) [reply src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.h:206]: I also found this anonymous struct strange and it basically means no struct being used. My original request to introduce a struct was to collect several categories of input parameters.

Comment (PDoakORNL) [reply src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.h:206]: The other issue here is the incoherence of encapsulation and the equivalence of the "engines" in this class but that is outside the scope of this PR.

File: src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.cpp
Comment (ye-luo) [reply src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.cpp]: Fixed

File: src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.h
Comment (ye-luo) [reply src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.h]: Name the struct as LMYOptions.

File: src/formic/utils/lmyengine/matrix_builder.h
Comment (PDoakORNL) [inline src/formic/utils/lmyengine/matrix_builder.h:225]: The is an abstraction to get the number of threads, see src/Concurrency/Info.hpp but the optimization has no guarantee it will run on that many threads so this just isn't an effective way to allocate concurrenct workspace.  It's possible this will work for awhile but it is at least a memory use bomb.

**Review by PDoakORNL — CHANGES_REQUESTED** [review]
Ok another iteration to go here.  Some of these broken Engines/EngineHandles and elements of formic could just be documented including todo's  But by the end of this series of PR's the illogical and/or wrong code needs to be gone.  Especially since sizable amounts of it appear to be new.

File: src/formic/utils/lmyengine/matrix_builder.h
Comment (ye-luo) [reply src/formic/utils/lmyengine/matrix_builder.h:225]: #4209 tracks the issue

File: src/formic/utils/lmyengine/engine.cpp
Comment (ye-luo) [reply src/formic/utils/lmyengine/engine.cpp]: Removed.

**Review by PDoakORNL — APPROVED** [review]

**Comment by ye-luo** [comment]
Test this please

**Pull request merged**
